### PR TITLE
Implement, test and document reindex

### PR DESCRIPTION
**Pull request by MazeChaZer**
Related issue #100

### CODE REVIEW

File: docs/functional-php.md [hunk 1]
```diff
@@ -277,6 +277,9 @@ is returned
 `mixed Functional\tail(array|Traversable $collection[, callable $callback])`
 Returns every element of the collection except the first one. Elements are optionally filtered by callback.
 
+`array Functional\reindex(array|Traversable $collection, callable $callback)`
+Reassigns all values to keys generated by a callback that is applied to every element.
```
Comment (lstrojny) [inline docs/functional-php.md]: "Reassign" sounds like mutation which we aren't doing

Comment (MazeChaZer) [inline docs/functional-php.md]: Right, I rewrote the description.

**Comment by lstrojny** [comment]
Looks good!

**Comment by lstrojny** [comment]
Thank you!

**Pull request merged**